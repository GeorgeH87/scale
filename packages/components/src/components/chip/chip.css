/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --font-size: var(--telekom-typography-font-size-body);
  --line-height: var(--telekom-typography-line-spacing-standard);
  --font-weight: var(--telekom-typography-font-weight-bold);
  --color-focus: var(--telekom-color-functional-focus);
  --focus-outline: var(--telekom-line-weight-highlight) solid
    var(--telekom-color-functional-focus);
  --box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.1), 0px 2px 8px rgba(0, 0, 0, 0.1);

  /*standard selected*/
  --background-standard-selected: var(--telekom-color-ui-extra-strong);
  --color-standard-selected: var(
    --telekom-color-text-and-icon-inverted-standard
  );
}
/**TODO: colors unexplained in figma  **/
.chip {
  border: var(--telekom-spacing-unit-x025) solid transparent;
  display: inline-flex;
  outline: none;
  padding: 0 var(--telekom-spacing-unit-x2);
  text-align: center;
  transition: all 0.15s ease-in-out;
  align-items: center;
  white-space: nowrap;
  border-radius: 1rem;
  border-color: rgba(255, 255, 255, 0.14);
  height: 32px;
  vertical-align: baseline;
  justify-content: center;
  cursor: default;
  background: var(--telekom-color-text-and-icon-inverted-standard);
  border: 1px solid var(--telekom-color-ui-border-standard);
}

.chip:not(.chip--disabled):focus {
  outline: var(--telekom-spacing-unit-x05) solid var(--color-focus);
  outline-offset: var(--telekom-spacing-unit-x025);
}

.chip:not(.chip--disabled):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered),
      var(--telekom-color-ui-state-fill-hovered)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

.chip:not(.chip--disabled):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed),
      var(--telekom-color-ui-state-fill-pressed)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

.chip--selected {
  background: var(--background-standard-selected);
  color: var(--color-standard-selected);
}

.chip--selected:not(.chip--disabled):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered-inverted),
      var(--telekom-color-ui-state-fill-hovered-inverted)
    ),
    var(--background-standard-selected);
}

.chip--selected:not(.chip--disabled):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed-inverted),
      var(--telekom-color-ui-state-fill-pressed-inverted)
    ),
    var(--background-standard-selected);
}

/**
label
**/

.chip-label {
  padding: 0 var(--telekom-spacing-unit-x3) 0 var(--telekom-spacing-unit-x3);
  font-weight: var(--telekom-typography-font-weight-bold);
  font-size: 1rem;
  line-height: 100%;
}

/**
icon front (slot)
**/

.chip slot[name='chip-icon']::slotted(*) {
  padding-left: var(--telekom-spacing-unit-x1);
  padding-top: var(--telekom-spacing-unit-x1);
}

/**
icon end
*/

.chip scale-icon-action-success {
  padding-right: var(--telekom-spacing-unit-x1);
}

.chip scale-icon-action-close {
  padding-right: var(--telekom-spacing-unit-x1);
}

/**
disabled
**/
.chip.chip--disabled {
  color: var(--telekom-color-text-and-icon-disabled);
  border: 1px solid var(--telekom-color-ui-border-disabled);
}

.chip--selected:not(.chip--type-outlined).chip--disabled {
  background: var(--telekom-color-ui-disabled);
  color: var(--telekom-color-text-and-icon-disabled);
  border: none;
}

.chip--type-outlined.chip--selected.chip--disabled {
  background-color: var(--telekom-color-text-and-icon-inverted-standard);
  color: var(--telekom-color-text-and-icon-disabled);
  border: 1px solid var(--telekom-color-ui-border-disabled);
}

/**
type outline 
**/

.chip.chip--type-outline.chip--selected {
  color: var(--telekom-color-text-and-icon-primary-standard);
  background: var(--telekom-color-text-and-icon-inverted-standard);
  border: 1px solid var(--telekom-color-text-and-icon-primary-standard);
}

.chip.chip--type-outline.chip--selected:not(.chip.chip--disabled):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered),
      var(--telekom-color-ui-state-fill-hovered)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
  color: var(--telekom-color-text-and-icon-primary-hovered);
  border: 1px solid var(--telekom-color-text-and-icon-primary-hovered);
}

.chip.chip--type-outline.chip--selected:not(.chip.chip--disabled):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed),
      var(--telekom-color-ui-state-fill-pressed)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
  color: var(--telekom-color-text-and-icon-primary-pressed);
  border: 1px solid var(--telekom-color-text-and-icon-primary-pressed);
}
