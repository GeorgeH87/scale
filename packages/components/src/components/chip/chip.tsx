/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, Prop, h, Host, Event, EventEmitter } from '@stencil/core';
import classNames from 'classnames';
import { emitEvent } from '../../utils/utils';
@Component({
  tag: 'scale-chip',
  styleUrl: './chip.css',
  shadow: true,
})
export class Chip {
  /** (optional) chip type */
  @Prop() type?: 'standard' | 'strong' | 'inversed' | 'colored' = 'standard';
  /** (optional) */
  @Prop() selected?: boolean = false;
  /** (optional) chip color */
  @Prop() color?:
    | 'cyan'
    | 'yellow'
    | 'green'
    | 'orange'
    | 'red'
    | 'violet'
    | 'brown'
    | 'olive'
    | 'teal'
    | 'black'
    | 'grey';
  /** (optional) chip href */
  @Prop() href?: string = '';
  /** (optional) chip dismissible */
  @Prop() dismissible?: boolean = false;
  /** (optional) chip label */
  @Prop() label?: string;
  /** (optional) chip target */
  @Prop() target?: string = '_self';
  /** (optional) chip disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** (optional) Change icon click event */
  @Event({ eventName: 'scale-change' }) scaleChange: EventEmitter<MouseEvent>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleChange' })
  scaleChangeLegacy: EventEmitter<MouseEvent>;
  /** (optional) Close icon click event */
  @Event({ eventName: 'scale-close' }) scaleClose: EventEmitter<MouseEvent>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleClose' })
  scaleCloseLegacy: EventEmitter<MouseEvent>;

  componentWillLoad() {
    if (this.dismissible) {
      this.selected = true;
    }
  }
  disconnectedCallback() {}

  handleClose = (event: MouseEvent) => {
    event.preventDefault();
    event.stopPropagation();
    if (this.disabled) {
      return;
    }
    emitEvent(this, 'scaleClose', event);
    console.log('close');
  };

  handleClick = (event: MouseEvent) => {
    if (this.dismissible) {
      this.handleClose(event);
      return;
    }
    event.preventDefault();
    event.stopPropagation();
    if (this.disabled) {
      return;
    }
    emitEvent(this, 'scaleChange', event);
    console.log('change');
  };

  getIcon() {
    if (this.dismissible) {
      return <scale-icon-action-close accessibility-title="close" size={16} />;
    } else if (!this.dismissible) {
      return (
        <scale-icon-action-success accessibility-title="success" size={16} />
      );
    }
  }

  render() {
    const Element = !!this.href && !this.disabled ? 'a' : 'span';
    const linkProps = !!this.href
      ? {
          href: this.href,
          target: this.target,
        }
      : {};

    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}

        <Element
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          {...linkProps}
          onClick={this.handleClick}
        >
          <slot name="left-icon" />
          <p class="chip-label">{this.label}</p>
          {this.selected ? this.getIcon() : null}
        </Element>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'chip';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      mode === 'basePart' ? 'base' : component,
      this.type && `${prefix}type-${this.type}`,
      this.color && `${prefix}color-${this.color}`,
      !!this.href && `${prefix}link`,
      !!this.disabled && `${prefix}disabled`
    );
  }
}
