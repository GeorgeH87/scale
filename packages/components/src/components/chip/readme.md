# scale-chip



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                    | Type                                | Default     |
| ------------- | ------------- | ------------------------------ | ----------------------------------- | ----------- |
| `color`       | `color`       | (optional) Chip color          | `"black" \| "default" \| "magenta"` | `'default'` |
| `disabled`    | `disabled`    | (optional) Chip disabled       | `boolean`                           | `false`     |
| `dismissable` | `dismissable` | (optional) Chip dismissable    | `boolean`                           | `false`     |
| `href`        | `href`        | (optional) Chip href           | `string`                            | `''`        |
| `iconOnly`    | `icon-only`   | (optional) Chip icon only      | `boolean`                           | `false`     |
| `styles`      | `styles`      | (optional) Injected CSS styles | `string`                            | `undefined` |
| `target`      | `target`      | (optional) Chip target         | `string`                            | `'_self'`   |


## Events

| Event         | Description                                                                                        | Type                      |
| ------------- | -------------------------------------------------------------------------------------------------- | ------------------------- |
| `scale-close` | (optional) Close icon click event                                                                  | `CustomEvent<MouseEvent>` |
| `scaleClose`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<MouseEvent>` |


## Dependencies

### Depends on

- [scale-icon-action-circle-close](../icons/action-circle-close)
- [scale-icon-device-mobile-services](../icons/device-mobile-services)

### Graph
```mermaid
graph TD;
  scale-chip --> scale-icon-action-circle-close
  scale-chip --> scale-icon-device-mobile-services
  style scale-chip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
