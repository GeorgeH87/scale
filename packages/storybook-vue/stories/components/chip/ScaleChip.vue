<template>
  <scale-chip
    :type="type"
    :selected="selected"
    :color="color"
    :href="href"
    :dismissible="dismissible"
    :label="label"
    :target="target"
    :disabled="disabled"
    :styles="styles"
    :icon-size="iconSize"
    @scaleChange="scaleChange"
    @scaleClose="scaleClose"
  >
    <slot name="chip-icon"></slot>
  </scale-chip>
</template>

<script>
import { action } from '@storybook/addon-actions';
export default {
  props: {
    type: {
      type: 'strong' | 'standard',
      default: 'standard',
    },
    selected: { type: Boolean, default: false },
    color: {
      type:
        'cyan' |
        'yellow' |
        'green' |
        'orange' |
        'red' |
        'violet' |
        'brown' |
        'olive' |
        'teal' |
        'grey',
    },
    href: { type: String, default: '' },
    dismissible: { type: Boolean, default: false },
    label: { type: String },
    target: { type: String, default: '_self' },
    disabled: { type: Boolean, default: false },
    styles: { type: String },
    iconSize: { type: Number, default: 16 },
  },
  methods: {
    scaleChange($event) {
      action('scaleChange');
      this.$emit('scaleChange', $event);
    },
    'scale-change'($event) {
      action('scale-change');
      this.$emit('scale-change', $event);
    },
    scaleClose($event) {
      action('scaleClose');
      this.$emit('scaleClose', $event);
    },
    'scale-close'($event) {
      action('scale-close');
      this.$emit('scale-close', $event);
    },
  },
};
</script>
