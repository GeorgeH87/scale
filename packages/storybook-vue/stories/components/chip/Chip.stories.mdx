import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs';
import ScaleChip from './ScaleChip.vue';
import { action } from '@storybook/addon-actions';

<Meta
  title="Beta Components/Chip"
  component={ScaleChip}
  argTypes={{
    label: {
      control: { type: 'text' },
    },
    type: {
      control: {
        type: 'select',
        options: ['standard', 'outline'],
      },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: {
    label: String,
    ...ScaleChip.props,
  },
  template: `
        <scale-chip
            :type="type"
            :selected="selected"
            :dismissible="dismissible"
            :label="label"
            :disabled="disabled"
            :styles="styles"
        >
        </scale-chip>
  `,
  methods: {
    action,
  },
});

export const TemplateIcon = (args, { argTypes }) => ({
  props: {
    label: String,
    ...ScaleChip.props,
  },
  template: `
    <scale-chip
      :selected="selected"
      :dismissible="dismissible"
      :label="label"
      :disabled="disabled"
      :styles="styles"
      >
      <div slot="chip-icon">
        <scale-icon-action-favorite accessibility-title="favorite" />
      </div>
    </scale-chip>
  `,
  methods: {
    action,
  },
});

<div
  style={{
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  }}
>
  <h1>Chip</h1>
  <img src="assets/beta.png" alt="Beta Component" />
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-chip label="Label"></scale-chip>
```

### Standard Selected

<Canvas withSource="none">
  <Story
    name="Standard Selected"
    args={{
      label: 'Label',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip label="Label" selected></scale-chip>
```

### Standard Disabled

<Canvas withSource="none">
  <Story
    name="Standard Disabled"
    args={{
      label: 'Label',
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled label="Label" selected></scale-chip>
```

### Standard Selected Disabled

<Canvas withSource="none">
  <Story
    name="Standard Selected Disabled"
    args={{
      label: 'Label',
      disabled: true,
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled label="Label" selected></scale-chip>
```

## Outline

<Canvas withSource="none">
  <Story
    name="Outline"
    args={{
      label: 'Label',
      type: 'outline',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip label="Label" type="outline"></scale-chip>
```

### Outline Selected

<Canvas withSource="none">
  <Story
    name="Outline Selected"
    args={{
      label: 'Label',
      type: 'outline',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip label="Label" selected></scale-chip>
```

### Outline Disabled

<Canvas withSource="none">
  <Story
    name="Outline Disabled"
    args={{
      type: 'outlined',
      label: 'Label',
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled type="outline" label="Label" selected></scale-chip>
```

### Outline Selected Disabled

<Canvas withSource="none">
  <Story
    name="Outline Selected Disabled"
    args={{
      type: 'outlined',
      label: 'Label',
      disabled: true,
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled type="outline" label="Label" selected></scale-chip>
```

## Dismissible

### Standard Dismissible

<Canvas withSource="none">
  <Story
    name="Standard Dismissible"
    args={{
      label: 'Label',
      dismissible: true,
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip dismissible selected label="Label"></scale-chip>
```

### Outline Dismissible

<Canvas withSource="none">
  <Story
    name="Outline Dismissible"
    args={{
      label: 'Label',
      dismissible: true,
      type: 'outline',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip dismissible selected label="Label" type="outline"></scale-chip>
```

## Icon + Text

<Canvas withSource="none">
  <Story
    name="Icon+Text"
    args={{
      label: 'Label',
    }}
  >
    {TemplateIcon.bind({})}
  </Story>
</Canvas>

```html
<scale-chip label="Label"
  ><div slot="chip-icon">
    <scale-icon-action-favorite accessibility-title="favorite" />
  </div>
</scale-chip>
```
