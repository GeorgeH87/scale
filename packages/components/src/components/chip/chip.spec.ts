import { newSpecPage } from '@stencil/core/testing';
import { Chip } from './chip';

describe('Chip', () => {
  let element;

  beforeEach(async () => {
    element = new Chip();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Chip],
      html: `<scale-chip label="Label Text"></scale-chip>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should handle css classes', () => {
    element.type = 'strong';
    expect(element.getCssClassMap()).toContain('chip--type-strong');

    element.color = 'green';
    expect(element.getCssClassMap()).toContain('chip--color-green');

    element.type = 'strong';
    expect(element.getCssClassMap()).toContain('chip--type-strong');

    element.selected = true;
    expect(element.getCssClassMap()).toContain('chip--selected');

    element.disabled = true;
    expect(element.getCssClassMap()).toContain('chip--disabled');

    element.href = 'http://example.com';
    expect(element.getCssClassMap()).toContain('chip--link');
  });
});
