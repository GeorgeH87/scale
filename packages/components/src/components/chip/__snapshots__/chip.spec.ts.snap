// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should match snapshot 1`] = `
<scale-chip label="Label Text">
  <mock:shadow-root>
    <span class="chip chip--type-standard" part="base type-standard">
      <slot name="chip-icon"></slot>
      <p class="chip-label">
        Label Text
      </p>
    </span>
  </mock:shadow-root>
</scale-chip>
`;
